fix(tui): resize at startup

muniter opened the pull request:
I've changed the PR of this approach to follow-on @erw7 https://github.com/neovim/neovim/issues/11330#issuecomment-1086617448

Which has two parts:

1. Removing the following code:

https://github.com/neovim/neovim/blob/af1b61f342618aa8d27c6e1dcbb0e360920ca89e/src/nvim/tui/tui.c#L1481-L1489

I debugged that this was the main issue in my system and perhaps in tiling window managers in general.

When the function `tui_guess_size` is called twice during startup the non-default 'columns' and 'lines' is usually true after the first time, because this same function already changed those values. Basically it's tricking itself, unrelated if the "lines" and "columns" options was set or not.

Added a test with the provided repro which fails before these changes.

2. Making got_winch a counter

If I understand things correctly this avoids resizing the host terminal until we have processed the last sigwinch, compared to the previous state of things:

> The second bug is that tui_grid_resize may change the size of the host terminal between the signal handler and the call to sigwinch_cb (note: the signal handle (uv__signal_handle) does not directly call the callback) when multiple SIGWINCH occur. I have confirmed that this actually happens in the following way using nvim with DLOG added.

Also removed the volatile since is not needed: https://github.com/neovim/neovim/pull/19210

PD: I'd be thrilled if @erw7 took over this, it's his proposed solution :D

Fixes #17285
Fixes #15044
Fixes #11330

zeertzjq commented on src/nvim/tui/tui.c:
```diff
@@ -496,7 +498,19 @@ static void sigcont_cb(SignalWatcher *watcher, int signum, void *data)
+  }
+}
+
+static inline void sigwinch_logic(void *data)
```
This name looks a bit strange. Maybe `handle_sigwinch` is better?

zeertzjq commented:
Hmm, this seems to cause a memory leak with `--version`.

zeertzjq commented:
This causes a memory leak:
```
./build/bin/nvim --clean +qa
```

bfredl commented on src/nvim/tui/tui.c:
```diff
@@ -508,6 +522,11 @@ static void sigwinch_cb(SignalWatcher *watcher, int signum, void *data)
   ui_schedule_refresh();
```
we should change this to only schedule a refresh if the TUI size actually changed (could be done in a follow-up PR if it gets too messy)

justinmk commented on src/nvim/tui/tui.c:
```diff
@@ -939,6 +962,15 @@ static void tui_grid_resize(UI *ui, Integer g, Integer width, Integer height)
   UGrid *grid = &data->grid;
   ugrid_resize(grid, (int)width, (int)height);
 
+  // When the UI starts, it might not be possible to resize correctly.
+  // https://github.com/neovim/neovim/issues/15044
```
```suggestion
  // When the UI starts, it might not be possible to resize correctly. #15044
```

justinmk commented on src/nvim/tui/tui.c:
```diff
@@ -135,6 +136,7 @@ typedef struct {
   char *space_buf;
 } TUIData;
 
+static bool volatile first_resize = true;
```
usual convention is "did" prefix:
```suggestion
static bool volatile did_resize = true;
```

muniter replied:
Went ahead an renamed, and also moved into scope where I think its purpose clearer, it also didn't need `volatile`.

Semantically `did_user_set_dimensions` is to not resize the **host terminal window**.

This PR and the new variable `did_first_resize` is about resizing the **TUI** correctly and **only at startup**.

zeertzjq commented on src/nvim/tui/tui.c:
```diff
@@ -474,6 +477,7 @@ static void tui_main(UIBridgeData *bridge, UI *ui)
   signal_watcher_stop(&data->cont_handle);
   signal_watcher_close(&data->cont_handle, NULL);
   signal_watcher_close(&data->winch_handle, NULL);
+  uv_close((uv_handle_t *)&data->initial_resize_timer, NULL);
```
Is is possible that `initial_resize_timer` can be uninitialized here?

muniter commented:
I think this is in a nice state. Haven't been able to test it. I had the following idea:

1. Launch nvim
2. Immediately block SIGWINCH using ffi (also tried `trap`).
3. Resize the host tty `job_resize`.
4. Check than nvim resized correctly.

But I'm failing at step 2. Wrote this code, nvim is still receiving SIGWINCH.

```lua
local ffi = require("ffi")
ffi.cdef[=[
typedef struct
{
  unsigned long int __val[(1024 / (8 * sizeof (unsigned long int)))];
} __sigset_t;
typedef __sigset_t sigset_t;
int sigaddset(sigset_t *set, int signo);
int sigprocmask(int how, const sigset_t *set, sigset_t *oset);
]=]

local defs = {
  SIG_BLOCK = 0,
  SIGWINCH = 28,
}

local mask_pt = ffi.new("sigset_t[1]")
local add = ffi.C.sigaddset(mask_pt[0], defs.SIGWINCH)
local mask = ffi.C.sigprocmask(defs.SIG_BLOCK, mask_pt[0], nil)
assert(add == 0 and mask == 0)  -- This is working
```

So I will run this build for a few days and report back.

CI failures seem unrelated, perhaps a re run?

erw7 commented:
I have checked the following steps and I do not believe this change will solve the problem.

```vim
nvim --clean
:call termopen(['nvim', '--clean', '--cmd', 'let start = reltime() | while v:true | if reltimefloat(reltime(start)) > 2 | break | endif | endwhile']) | slee
p 500m | vs
```

erw7 commented:
The cause and fix for the bug you attempted to fix in this PR is described by me in https://github.com/neovim/neovim/issues/11330#issuecomment-1086617448. Unfortunately, I don't think this PR method will fix it.

muniter committed:
fix(tui): resize at startup

The deleted code is not needed after 402b4e8

It caused the condition to false positive when the function was called
more than once before startup (first normal mode). Being itself what set
the dimension and not the user, locking the size of the screen to an
incorrect size.

Make got_winch an int to mitigate: tui_grid_resize changing the size of
the host terminal between the signal handler and the call to
sigwinch_cb. Since the actual signal handler uv__signal_handle doesn't
directly call the callback, the event loop does.

muniter commented:
I've updated the PR description, with the change of direction.

justinmk commented on test/functional/terminal/tui_spec.lua:
```diff
@@ -85,6 +85,23 @@ describe('TUI', function()
 
+  it('works correctly at startup', function()
+    screen:try_resize(50, 10)
+    feed_command([[call termopen(['build/bin/nvim', '--clean', '--cmd', 'let start = reltime() | while v:true | if reltimefloat(reltime(start)) > 2 | break | endif | endwhile']) | sleep 500m | vs new]])
```
```suggestion
    feed_command([[call termopen([v:progpath, '--clean', '--cmd', 'let start = reltime() | while v:true | if reltimefloat(reltime(start)) > 2 | break | endif | endwhile']) | sleep 500m | vs new]])
```

justinmk commented on test/functional/terminal/tui_spec.lua:
```diff
@@ -85,6 +85,23 @@ describe('TUI', function()
     assert_alive()
   end)
 
+  it('works correctly at startup', function()
```
"works correctly" is almost always unnecessary in a test description, because tests are usually testing that things work correctly.
```suggestion
  it('resize at startup', function()
```

justinmk commented on test/functional/terminal/tui_spec.lua:
```diff
@@ -85,6 +85,23 @@ describe('TUI', function()
+      {4:~                        }│{4:~                       }|
+      {4:~                        }│{5:[No Name]   0,0-1    All}|
+      {4:~                        }│                        |
+      {5:new                       }{1:</neovim/build/bin/nvim }|
```
can use `{MATCH:x}` so that we don't hardcode the path, see `screen.lua`.

```suggestion
      {5:new                       }{MATCH:<.*[/\]nvim }|
```

justinmk commented:
Very eager to get this fixed! Has a minor conflict because of https://github.com/neovim/neovim/commit/436747752bf1016d93101de24e39e192a0b721b4

muniter committed:
fix(test): screen.lua nil index

When actual_rows and expected_rows are different avoid a nil index

muniter commented:
Found a bug in screen.lua that was reproduced by the test in this PR when using `MATCH`. 
Added it here since it was such a small change and if I move it to another PR this one has to be rebased on top of it. But I'm fine moving it if needed.

justinmk commented:
Oh, we need to update the doc here: https://github.com/neovim/neovim/issues/17285#issuecomment-1030512165

muniter commented:
> Do we need to update the doc here?: [#17285 (comment)](https://github.com/neovim/neovim/issues/17285#issuecomment-1030512165)

I don't think so, the code indicates we follow that process. From my point of view we  actually removed an undocumented case 🤔
